### DISCUSSION TITLE
Name "Scene" is ambiguous

### DISCUSSION THREAD

**Discussion by PhoebeMitchell** [Studio-Lovelies/GG-JointJustice-Unity#236] [General]
Following on from the discussion here:
https://github.com/Studio-Lovelies/GG-JointJustice-Unity/issues/234

The action "&SCENE" is ambiguous because it could refer to either Unity scenes, or our BGScenes.

Should the naming for the SCENE action and BGScenes be updated?

**Comment by PhoebeMitchell** [comment]
> > There currently isn't a LOAD_SCENE action but I agree that would be a good idea. Then as you say LOAD_SCENE could load the credits, or minigames, etc.
> 
> Exactly! Basically for anything, where we exit the narrative loop.
> 
> > LOAD_SCRIPT doesn't reload the game scene, just sets up the next narrative script, so I think a separate LOAD_SCENE action would be better.
> 
> Absolutely makes sense; the term "reuse" was a bit irritating, I admit. I actually was referring to creating a new `LOAD_SCENE` action, referring to loading Unity Scenes, (rather than the current narrative scenes).
> 
> > SCENE just enables and disables BGScenes and I think it would be better to make the credits its own scene within Unity.
> 
> Maybe similarly to the overlap with the term "script" (referring to either an Ink Script or C# Script), we could come up with something to use instead, that doesn't clash with existing Unity vocabulary? That way we wouldn't have to qualify them as Ink Scenes and Unity Scenes. I could see something like "stage", "act" or "location" working for scene. "Screenplay" could be an alternative for Ink scripts.

Yeah, there's a little ambiguity regarding BGScenes and Unity scenes. I guess `&BGSCENE` could work but tbh I don't like that name much either. Location works. Stage too. Possibly "view"? Act implies a section of a story to me so not so much that one.

I think "Ink script" is pretty unambiguous. It's a script that is an ink file, and they aren't used directly in game so naming shouldn't be much of a problem.

**Reply by ViMaSter** [reply]
> Yeah, there's a little ambiguity regarding BGScenes and Unity scenes. I guess &BGSCENE could work but tbh I don't like that name much either.

I can absolutely get behind that and agree with you on the `&BGSCENE` assessment too! My main issue with it is that both objects don't share the same underlying data in any way, but that's what the current naming implies to me. My mind immediately goes to  
`class BGScene : Scene`  
`class UnityScene : Scene`  
except that's not what happens underneath. : D

> Location works. Stage too. Possibly "view"? Act implies a section of a story to me so not so much that one.

`Stage` seems like the one that has the least associations outside of it, which makes it my personal favorite. "View" could be tied to the camera or MVC, "location" might relate to physics/UI.

---

> I think "Ink script" is pretty unambiguous. It's a script that is an ink file, and they aren't used directly in game so naming shouldn't be much of a problem.

I definitely agree with that! My main gripe is with the term "scripts" all on its own, which right now can now mean two (in my mind at least) very different things:
C# scripts in Unity: Compiled before the game is run, defining the game's feature set because they can interact with the `UnityEngine` and `UnityEditor` assemblies and exist in every part of the game.   
Ink scripts: Parsed only during runtime, limited to a specific subset of instructions, and only define the flow of narrative gameplay.

With something like "screenplays" and "scripts", the ambiguity of the word "scripts" would be gone. Right now we need context to know whether it's "C# scripts" or "Ink scripts" we are talking about, which isn't always the case. (For example, we have the folders `Assets/Scripts` or `Assets/Tests/PlayModeTests/Scripts` which are actually limited to C# scripts; not sure how Unity would handle special characters, if we renamed them to `Assets/C# Scripts` and `Assets/Tests/PlayModeTests/C# Scripts` respectively.)

That said, if everyone is okay with continuing to just call them "C# scripts" or "Ink scripts", I can also see that work, but would like to enforce that separation a bit more strictly moving forward. : D

**Reply by PhoebeMitchell** [reply]
It hasn't proved to me much of an issue so far. That said some consistency in naming wouldn't hurt.

Originally we settled on calling the ink scripts them narrative scripts, which is how they're referred to in the game's code. There could be an argument made that a narrative script is actually a container for an ink story (an ink story being an object created from a compiled ink script in JSON format). There's a bit of a process between writing an ink script and having it be playable in game (Ink file->JSON file->Ink story->NarrativeScript). The directory where the Ink scripts are stored is called InkDialogueScripts. DialogueScripts is what they were called in the Lua build I think, though in our build "Dialogue" refers to a specific mode in game.

> For example, we have the folders Assets/Scripts or Assets/Tests/PlayModeTests/Scripts which are actually limited to C# scripts

Don't forget "TestScripts" which contains just Ink and JSON files :p

**Reply by PhoebeMitchell** [reply]
Oh, and I'm fine with the name "Stage" for BGScenes.

**Reply by ViMaSter** [reply]
Ooo~, `Story` is a really good term. Could we just go for that? 👀 

Summarized we would have...
- Scripts: C# (editor and game) files
- Stories: Ink files
- Stage: A background image with actor slots

That would mean `TestScripts` would become `TestStories`. Overall, this feels pretty intuitive to me.

**Reply by PhoebeMitchell** [reply]
That works. Would NarrativeScript stay as it is?
Also we'd have to make sure nobody's confusing "stories" with the "stories" on our issues page, which aren't actually stories but are called that sometimes.

**Reply by ViMaSter** [reply]
> That works. Would NarrativeScript stay as it is?

Nice!  
I think I would change this to `Story` as well, realizing full well we've gone full circle, but at least there's a proper definition for what a Story is now. : D
(I'm going to add a "Vocabulary" page to the Wiki with names, what they mean, and why they exist. Cause I know _I_ will forget, so I can't ask others to remember.)

`NarrativeScriptPlaylist` would become `StoryPlaylist`  
`NarrativeScriptTests` would become `StoryTests`

> Also we'd have to make sure nobody's confusing "stories" with the "stories" on our issues page, which aren't actually stories but are called that sometimes.

The first professional project I've worked on had the classic `Bugs`/`Tasks`/`User Stories`/`Epic` structure (in which a  
`User Story` follows _"As `X` I'm able to `Y` in order to `Z`"_ and **isn't actionable in itself** and  
`Tasks` are "**actionable** while **relevant only to the department working on them** to make the User Story possible"), so if I had to pick one of them to represent GitHub Issues, I'd probably go with `Tasks` over "`(User) Stories`". But since they're named Issues, I think it's a good idea to just stick with that! : D

**Reply by PhoebeMitchell** [reply]
NarrativeScript cannot be called Story, that name is already taken by Story in Ink.
NarrativeScripts purpose is to store all the things related to a story (i.e., the story itself, its text asset (currently called "Script" lol), the objects in that story (along with reading the story and loading objects). What would you call something that does all that?

**Reply by ViMaSter** [reply]
Ah, right, forgot about that, thanks for the reminder! I could see something like `StoryData(Container)`/`StoryAssets(Container)` working? Or do you find that to be too broad of a term?

**Reply by PhoebeMitchell** [reply]
StoryData works I think
